hazelcast/src/main/java/com/hazelcast/scheduledexecutor/impl/ScheduledExecutorContainer.java

```diff
@@ -0,0 +1,426 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.scheduledexecutor.impl;
+
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.scheduledexecutor.DuplicateTaskException;
+import com.hazelcast.scheduledexecutor.ScheduledTaskHandler;
+import com.hazelcast.scheduledexecutor.ScheduledTaskStatistics;
+import com.hazelcast.scheduledexecutor.StaleTaskException;
+import com.hazelcast.scheduledexecutor.StatefulTask;
+import com.hazelcast.scheduledexecutor.impl.operations.ResultReadyNotifyOperation;
+import com.hazelcast.scheduledexecutor.impl.operations.SyncStateOperation;
+import com.hazelcast.spi.InvocationBuilder;
+import com.hazelcast.spi.NodeEngine;
+import com.hazelcast.spi.Operation;
+import com.hazelcast.spi.OperationService;
+import com.hazelcast.spi.impl.executionservice.InternalExecutionService;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.concurrent.Callable;
+import java.util.concurrent.Delayed;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static com.hazelcast.scheduledexecutor.impl.DistributedScheduledExecutorService.SERVICE_NAME;
+import static com.hazelcast.util.ExceptionUtil.sneakyThrow;
+
+public class ScheduledExecutorContainer {
+
+    private final ILogger logger;
+
+    private final String name;
+
+    private final NodeEngine nodeEngine;
+
+    private final InternalExecutionService executionService;
+
+    private final int partitionId;
+
+    private final int durability;
+
+    private final Map<String, ScheduledTaskDescriptor> tasks;
+
+    private final Map<String, BackupTaskDescriptor> backups;
+
+    public ScheduledExecutorContainer(String name, int partitionId, NodeEngine nodeEngine,
+                                      int durability) {
+        this(name, partitionId, nodeEngine, durability, new HashMap<String, BackupTaskDescriptor>());
+    }
+
+    public ScheduledExecutorContainer(String name, int partitionId, NodeEngine nodeEngine,
+                                      int durability, Map<String, BackupTaskDescriptor> backups) {
+        this.logger = nodeEngine.getLogger(getClass());
+        this.name = name;
+        this.nodeEngine = nodeEngine;
+        this.executionService = (InternalExecutionService) nodeEngine.getExecutionService();
+        this.partitionId = partitionId;
+        this.durability = durability;
+        this.tasks = new HashMap<String, ScheduledTaskDescriptor>();
+        this.backups = backups;
+    }
+
+    public <V> ScheduledFuture<V> schedule(TaskDefinition definition) {
+        checkNotDuplicateTask(definition.getName());
```

Comment:
Looks like this `member-partition` will cause more problem. Imagine 2 clients try to schedule a named task on a specific member at the same time. They both send `ScheduleTaskOperation`, they run on generic threads in parallel. They do the name checking at the same time and pass.